src/plugins/intel_cpu/src/nodes/multinomial.hpp

```diff
@@ -0,0 +1,189 @@
+        std::vector<P> m_max_per_batch(m_batches_count);
+        std::vector<P> m_random_samples(m_output_elements_count);
+
+        // exp & cumsum
+        if (m_log_probs) {
+            parallel_for(m_batches_count, [&](size_t idx) {
+                auto start_idx = idx * m_probs_count;
+                m_cdf[start_idx] = std::exp(probs[start_idx]);
+                for (size_t prev = start_idx, curr = prev + 1; curr < (start_idx + m_probs_count); ++prev, ++curr) {
+                    m_cdf[curr] = std::exp(probs[curr]) + m_cdf[prev];
+                }
+            });
+        } else {
+            parallel_for(m_batches_count, [&](size_t idx) {
+                auto start_idx = idx * m_probs_count;
+                m_cdf[start_idx] = probs[start_idx];
+                for (size_t prev = start_idx, curr = prev + 1; curr < (start_idx + m_probs_count); ++prev, ++curr) {
+                    m_cdf[curr] = probs[curr] + m_cdf[prev];
+                }
+            });
+        }
+
+        // TODO RandomUniform - should use RandomUniform kernel to match other frameworks' seed results
+        uint64_t seed;
+        if (m_global_seed == 0 && m_op_seed == 0) {
+            seed = std::time(NULL);
+        } else {
+            std::srand(m_global_seed);
+            std::srand(m_op_seed + std::rand());
+            seed = std::rand();
+        }
+
+        std::mt19937 gen(seed);
+        std::uniform_real_distribution<float> dist(0.0f, 1.0f);
+        for (size_t idx = 0lu; idx < m_output_elements_count; ++idx) {
+            m_random_samples[idx] = static_cast<P>(dist(gen));
+        }
+
+        // max & divide
+        const auto min_value_of_max = std::numeric_limits<P>::min();
+        parallel_for(m_batches_count, [&](size_t idx) {
+            m_max_per_batch[idx] = std::max(m_cdf[(idx + 1) * m_probs_count - 1], min_value_of_max);
+        });
+        parallel_for(m_input_elements_count, [&](size_t idx) {
+            size_t idx_max_elem = idx / m_probs_count;
+            m_cdf[idx] /= m_max_per_batch[idx_max_elem];
+        });
+
+        if (m_with_replacement) {
+            parallel_for(m_batches_samples_probs_count, [&](size_t idx) {
+                size_t idx_batch = idx / m_samples_probs_count;
+                size_t idx_num_samples_probs = idx % m_samples_probs_count;
+                size_t idx_prob = idx_num_samples_probs % m_probs_count;
+                size_t idx_sample = idx_num_samples_probs / m_probs_count;
+
+                size_t idx_input = idx_batch * m_probs_count + idx_prob;
+                size_t idx_output = idx_batch * m_samples_count + idx_sample;
+                if (m_random_samples[idx_output] <= m_cdf[idx_input] &&
+                    (!idx_prob || m_random_samples[idx_output] > m_cdf[idx_input - 1])) {
+                    output[idx_output] = static_cast<O>(idx_prob);
+                }
+            });
+        } else {
+            parallel_for(m_batches_count, [&](size_t idx_batch) {
```

Comment:
parallel_for2d

Reply:
Cannot be applied here - when sampling without replacement, every time a new value is sampled from a batch, the CDF must be modified before the next sample operation is permitted, hence we cannot use parallelism in that case, only sequential